### PR TITLE
Support scan for Trace_ELBO

### PR DESCRIPTION
**Pull request by deoxyribose**
Issue #1685

Reuse the `substitute_stack` to store the replay trace, and replay the sites in the scan'ed function one iteration at a time.

### CODE REVIEW

**Comment by fehiepsi** [comment]
Nice support, @deoxyribose! Could you add tests for this change?

**Comment by deoxyribose** [comment]
I've added a test that does SVI with an AutoNormal guide, and checks that results are relatively accurate. I've kept model and data size to a minimum, the test takes around 3-4 seconds on my machine. There already was a test combining scan, SVI and AutoNormal (test_subsample_guide in test/infer/test_autoguide.py), but it doesn't run inference.

File: numpyro/contrib/control_flow/scan.py [hunk 1]
```diff
@@ -15,6 +15,23 @@
...
+            site["name"] not in trace
+            or site_len != length
+            or site["type"] not in ("sample", "deterministic")
+        ):
+            return site
+
+        site["value"] = site["value"][i]
+        return site
+
+    return {k: get_ith_value(v.copy()) for k, v in replay_trace.items()}
```
Comment (fehiepsi) [inline numpyro/contrib/control_flow/scan.py:32]: nit: it is better to do the copy inside `git_ith_value`

**Review by fehiepsi — APPROVED** [review]
Looks great! Thanks, @deoxyribose !

**Pull request merged**